builtin/v11/migration/top.go

```diff
@@ -0,0 +1,120 @@
+package migration
+
+import (
+	"context"
+
+	system10 "github.com/filecoin-project/go-state-types/builtin/v10/system"
+	adt10 "github.com/filecoin-project/go-state-types/builtin/v10/util/adt"
+
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/builtin"
+	"github.com/filecoin-project/go-state-types/manifest"
+	"github.com/filecoin-project/go-state-types/migration"
+
+	"github.com/ipfs/go-cid"
+	cbor "github.com/ipfs/go-ipld-cbor"
+	"golang.org/x/xerrors"
+)
+
+// MigrateStateTree Migrates the filecoin state tree starting from the global state tree and upgrading all actor state.
+// The store must support concurrent writes (even if the configured worker count is 1).
+func MigrateStateTree(ctx context.Context, store cbor.IpldStore, newManifestCID cid.Cid, actorsRootIn cid.Cid, priorEpoch abi.ChainEpoch, cfg migration.Config, log migration.Logger, cache migration.MigrationCache) (cid.Cid, error) {
+	if cfg.MaxWorkers <= 0 {
+		return cid.Undef, xerrors.Errorf("invalid migration config with %d workers", cfg.MaxWorkers)
+	}
+
+	adtStore := adt10.WrapStore(ctx, store)
+
+	// Load input and output state trees
+	actorsIn, err := builtin.LoadTree(adtStore, actorsRootIn)
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("loading state tree: %w", err)
+	}
+
+	// load old manifest data
+	systemActor, ok, err := actorsIn.GetActorV5(builtin.SystemActorAddr)
+	if err != nil {
+		return cid.Undef, xerrors.Errorf("failed to get system actor: %w", err)
+	}
+
+	if !ok {
+		return cid.Undef, xerrors.New("didn't find system actor")
+	}
+
+	var systemState system10.State
+	if err := store.Get(ctx, systemActor.Head, &systemState); err != nil {
+		return cid.Undef, xerrors.Errorf("failed to get system actor state: %w", err)
+	}
+
+	var oldManifestData manifest.ManifestData
+	if err := store.Get(ctx, systemState.BuiltinActors, &oldManifestData); err != nil {
+		return cid.Undef, xerrors.Errorf("failed to get old manifest data: %w", err)
+	}
+
+	// load new manifest
+	var newManifest manifest.Manifest
+	if err := adtStore.Get(ctx, newManifestCID, &newManifest); err != nil {
+		return cid.Undef, xerrors.Errorf("error reading actor manifest: %w", err)
+	}
+
+	if err := newManifest.Load(ctx, adtStore); err != nil {
+		return cid.Undef, xerrors.Errorf("error loading actor manifest: %w", err)
+	}
+
+	// Maps prior version code CIDs to migration functions.
+	migrations := make(map[cid.Cid]migration.ActorMigration)
+	// Set of prior version code CIDs for actors to defer during iteration, for explicit migration afterwards.
+	deferredCodeIDs := make(map[cid.Cid]struct{})
+
+	miner10Cid := cid.Undef
+	for _, oldEntry := range oldManifestData.Entries {
+		if oldEntry.Name == manifest.MinerKey {
```

Comment:
nit: No correctness problem since you overwrite below, but to be consistent and clear I think you should do the same thing with oldEntry.Name == manifest.SystemKey in another if case branch OR remove this if statement and just overwrite the miner migration the way you overwrite the system migration.